main/main.go

```diff
@@ -254,76 +265,178 @@ func main() {
+	manualWarpMessages []*relayerTypes.WarpMessageInfo,
 	cfg *config.Config,
 ) error {
-	logger.Info(
-		"Creating relayer",
-		zap.String("originBlockchainID", sourceSubnetInfo.BlockchainID),
+	// Dial the eth client
+	ethClient, err := ethclient.DialWithConfig(
+		context.Background(),
+		sourceBlockchain.RPCEndpoint.BaseURL,
+		sourceBlockchain.RPCEndpoint.HTTPHeaders,
+		sourceBlockchain.RPCEndpoint.QueryParams,
 	)
+	if err != nil {
+		logger.Error(
+			"Failed to connect to node via RPC",
+			zap.String("blockchainID", sourceBlockchain.BlockchainID),
+			zap.Error(err),
+		)
+		return err
+	}
 
-	listener, err := relayer.NewListener(
+	// Create the ApplicationRelayers
+	applicationRelayers, minHeight, err := createApplicationRelayers(
+		ctx,
 		logger,
 		metrics,
 		db,
 		ticker,
-		sourceSubnetInfo,
+		sourceBlockchain,
 		network,
-		responseChan,
-		destinationClients,
 		messageCreator,
+		cfg,
+		ethClient,
+		destinationClients,
+	)
+	if err != nil {
+		logger.Error(
+			"Failed to create application relayers",
+			zap.String("blockchainID", sourceBlockchain.BlockchainID),
+			zap.Error(err),
+		)
+		return err
+	}
+	logger.Info(
+		"Created application relayers",
+		zap.String("blockchainID", sourceBlockchain.BlockchainID),
+	)
+
+	// Create the Listener
+	listener, err := relayer.NewListener(
+		logger,
+		sourceBlockchain,
 		relayerHealth,
 		cfg,
+		applicationRelayers,
+		minHeight,
+		ethClient,
 	)
 	if err != nil {
 		return fmt.Errorf("failed to create listener instance: %w", err)
 	}
 	logger.Info(
 		"Created listener",
-		zap.String("blockchainID", sourceSubnetInfo.BlockchainID),
+		zap.String("blockchainID", sourceBlockchain.BlockchainID),
 	)
-
-	// Send any messages that were specified in the configuration
-	for _, warpMessage := range manualWarpMessages {
-		logger.Info(
-			"Relaying manual Warp message",
-			zap.String("blockchainID", sourceSubnetInfo.BlockchainID),
-			zap.String("warpMessageBytes", hex.EncodeToString(warpMessage.UnsignedMsgBytes)),
+	err = listener.ProcessManualWarpMessages(logger, manualWarpMessages, sourceBlockchain)
+	if err != nil {
+		logger.Error(
+			"Failed to process manual Warp messages",
+			zap.String("blockchainID", sourceBlockchain.BlockchainID),
+			zap.Error(err),
 		)
-		err := listener.RouteManualWarpMessage(warpMessage)
-		if err != nil {
-			logger.Error(
-				"Failed to relay manual Warp message. Continuing.",
-				zap.Error(err),
-				zap.String("warpMessageBytes", hex.EncodeToString(warpMessage.UnsignedMsgBytes)),
-			)
-			continue
-		}
 	}
 
 	logger.Info(
 		"Listener initialized. Listening for messages to relay.",
-		zap.String("originBlockchainID", sourceSubnetInfo.BlockchainID),
+		zap.String("originBlockchainID", sourceBlockchain.BlockchainID),
 	)
 
 	// Wait for logs from the subscribed node
 	// Will only return on error or context cancellation
 	return listener.ProcessLogs(ctx)
 }
 
+func createApplicationRelayers(
```

Comment:
Can we add a doc comment about this function and its intended use? I think it creates the application relayers for a specific source subnet, in which case may want to rename to make that clear.